presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +503,189 @@ else if (!result.getBlocked().isDone()) {
+                .collect(Collectors.joining(", ")));
+        return section;
+    }
+
+    private void performRuntimeOptimizations(StreamingSubPlan subPlan, Map<PlanFragment, PlanFragment> oldToNewFragmentMapping)
+    {
+        PlanFragment fragment = subPlan.getFragment();
+        PlanNode newRoot = fragment.getRoot();
+        for (PlanOptimizer optimizer : runtimePlanOptimizers) {
+            newRoot = optimizer.optimize(newRoot, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
+        }
+        // The partitioningScheme should stay the same
+        // even if the root's outputVariable layout is changed.
+        if (newRoot != fragment.getRoot()) {
+            oldToNewFragmentMapping.put(fragment, new PlanFragment(
+                    fragment.getId(),
+                    newRoot,
+                    fragment.getVariables(),
+                    fragment.getPartitioning(),
+                    scheduleOrder(newRoot),
+                    fragment.getPartitioningScheme(),
+                    fragment.getStageExecutionDescriptor(),
+                    fragment.isOutputTableWriterFragment(),
+                    fragment.getStatsAndCosts(),
+                    fragment.getJsonRepresentation()));
+        }
+    }
+
+    /**
+     * Utility function that rebuild a StreamingPlanSection, re-create stageExecutionAndScheduler for each of its stage, and finally update the stageExecutions map.
+     */
+    private void updateStageExecutions(StreamingPlanSection section, Map<PlanFragment, PlanFragment> oldToNewFragment)
+    {
+        StreamingPlanSection newSection = new StreamingPlanSection(rewriteStreamingSubPlan(section.getPlan(), oldToNewFragment), section.getChildren());
+        PlanFragment sectionRootFragment = newSection.getPlan().getFragment();
+        Optional<int[]> bucketToPartition;
+        OutputBuffers outputBuffers;
+        ExchangeLocationsConsumer locationsConsumer;
+        if (isRootFragment(sectionRootFragment)) {
+            bucketToPartition = Optional.of(new int[1]);
+            outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                    .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                    .withNoMoreBufferIds();
+            OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                    updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+        }
+        else {
+            bucketToPartition = Optional.empty();
+            outputBuffers = createDiscardingOutputBuffers();
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+        }
+        SectionExecution sectionExecution = sectionExecutionFactory.createSectionExecutions(
```

Comment:
@rschlussel : Is `SectionExecutionFactory` introduced for the new sql query scheduler? So essentially we are calling something in the new scheduler from the legacy scheduler? ;)

Reply:
sectionexecutionfactory is already used in the legacySqlQueryScheduler (and the logic was originally extracted from there)